cpp/src/text/edit_distance.cu

```diff
@@ -108,16 +94,179 @@ struct calculate_compute_buffer_fn {
+ *  S  4  z  ↘  .  .  .  .
+ *  T  5  →  .  .  .  .  X
+ *
+ * Computing a diagonal only requires the previous 2 and each value
+ * can be computed independently/parallel a tile at a time.
+ *
+ * The diagonal approach inspired by:
+ * https://ashvardanian.com/posts/stringwars-on-gpus/#dynamic-programming-and-levenshtein-evaluation-order
  */
+template <typename Iterator, int32_t tile_size = cudf::detail::warp_size>
+__device__ void compute_distance(Iterator input1,
+                                 cudf::size_type length1,
+                                 Iterator input2,
+                                 cudf::size_type length2,
+                                 cudf::size_type* d_buffer,
+                                 cudf::size_type* d_result)
+{
+  namespace cg        = cooperative_groups;
+  auto const block    = cg::this_thread_block();
+  auto const tile     = cg::tiled_partition<tile_size>(block);
+  auto const lane_idx = tile.thread_rank();
+
+  // shortcut if one of the strings is empty
+  // (null rows are mapped here as well)
+  if (length1 == 0 || length2 == 0) {
+    if (lane_idx == 0) { *d_result = length1 == 0 ? length2 : length1; }
+    return;
+  }
+
+  // setup the 3 working vectors for this string
+  auto v0 = d_buffer;
+  auto v1 = v0 + length1 + 2;
+  auto v2 = v1 + length1 + 2;
+  if (lane_idx == 0) {
+    v0[0] = 0;  // first diagonal
+    v1[0] = 1;  // second diagonal
+    v1[1] = 1;
+  }
+
+  // utility for navigating the diagonal of the matrix of characters for the 2 strings
+  auto next_itr = [](Iterator sitr, cudf::size_type length, Iterator itr, cudf::size_type offset) {
+    if constexpr (cuda::std::is_pointer_v<Iterator>) {
+      itr = sitr - offset;  // ASCII iterator
+    } else {
+      auto const pos = sitr.position() - offset;  // minimizes character counting
+      itr += (pos >= 0) && (pos < length) ? (pos - itr.position()) : 0;
+    }
+    return itr;
+  };
+
+  // top-left of the matrix
+  // includes the diagonal one after the max(length1,length2) diagonal
+  for (auto idx = 0; idx < length1; ++idx, ++input1) {
+    auto const n = idx + 2;      // diagonal length
+    auto const a = n > length1;  // extra diagonal adjust
+
+    auto jdx = static_cast<cudf::size_type>(lane_idx);
+    auto it1 = input1;
+    auto it2 = input2;
+
+    auto tile_count = cudf::util::div_rounding_up_safe(n + 1, tile_size);
+    while (tile_count--) {
+      auto const offset = (jdx - 1);
+      // locate the 2 characters to compare along the diagonal
+      it1 = next_itr(input1, length1, it1, offset);
+      it2 = next_itr(input2, length2, it2, -offset);
+      if (jdx == 0) {
+        if (!a) { v2[0] = n; }
+      } else if (jdx < n) {
+        auto const sc = v0[jdx - 1] + (*it1 != *it2);
+        auto const dc = v1[jdx - 1] + 1;
+        auto const ic = v1[jdx] + 1;
+        v2[jdx - a]   = cuda::std::min(cuda::std::min(sc, dc), ic);
+      } else if (jdx == n) {
+        v2[n - a] = n;
+      }
```

Comment:
To make sure I understand the parallelization scheme you're applying here, the idea is that we use `tile_size` threads to update elements of v2 in parallel. If the size of v2 (i.e. the size of the string) is less than `tile_size`, then at each iteration we'll have `tile_size - length1` idle threads, right? Conversely if `length1 > tile_size` then we step through v2 in blocks of size `tile_size`, which is the `while(tile_count--)` loop?

Reply:
I guess we're not really aiming for peak perf here so it's OK if we don't think about this right now, but maybe there's some opportunity for work-stealing to make this more efficient?